frame/staking/src/lib.rs

```diff
@@ -2919,50 +2959,36 @@ impl<T: Config> Pallet<T> {
 
 	/// Get all of the voters that are eligible for the npos election.
 	///
+	/// `voter_count` imposes an implicit cap on the number of voters returned; care should be taken
+	/// to ensure that it is accurate.
+	///
 	/// This will use all on-chain nominators, and all the validators will inject a self vote.
 	///
 	/// ### Slashing
 	///
 	/// All nominations that have been submitted before the last non-zero slash of the validator are
 	/// auto-chilled.
 	///
-	/// Note that this is VERY expensive. Use with care.
-	pub fn get_npos_voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		let weight_of = Self::slashable_balance_of_fn();
-		let mut all_voters = Vec::new();
-
-		for (validator, _) in <Validators<T>>::iter() {
-			// Append self vote.
-			let self_vote = (validator.clone(), weight_of(&validator), vec![validator.clone()]);
-			all_voters.push(self_vote);
-		}
+	/// Note that this is fairly expensive: it must iterate over the min of `maybe_max_len` and
+	/// `voter_count` voters. Use with care.
+	pub fn get_npos_voters(
+		maybe_max_len: Option<usize>,
+		voter_count: usize,
+	) -> Vec<VotingDataOf<T>> {
+		let wanted_voters = maybe_max_len.unwrap_or(voter_count).min(voter_count);
 
-		// Collect all slashing spans into a BTreeMap for further queries.
+		let weight_of = Self::slashable_balance_of_fn();
+		// collect all slashing spans into a BTreeMap for further queries.
 		let slashing_spans = <SlashingSpans<T>>::iter().collect::<BTreeMap<_, _>>();
 
-		for (nominator, nominations) in Nominators::<T>::iter() {
-			let Nominations { submitted_in, mut targets, suppressed: _ } = nominations;
-
-			// Filter out nomination targets which were nominated before the most recent
-			// slashing span.
-			targets.retain(|stash| {
-				slashing_spans
-					.get(stash)
-					.map_or(true, |spans| submitted_in >= spans.last_nonzero_slash())
-			});
-
-			if !targets.is_empty() {
-				let vote_weight = weight_of(&nominator);
-				all_voters.push((nominator, vote_weight, targets))
-			}
-		}
-
-		all_voters
+		VoterList::<T>::iter()
+			.filter_map(|node| node.voting_data(&weight_of, &slashing_spans))
+			.take(wanted_voters)
+			.collect()
 	}
 
-	/// This is a very expensive function and result should be cached versus being called multiple times.
```

Comment:
This comment seems relevant still

Reply:
You're right. Good catch! https://github.com/paritytech/substrate/pull/9081/commits/033d87faec30339328f87d2ec7abeb67bb3a3e42.